test-framework/src/main/java/org/apache/brooklyn/test/framework/TestFrameworkAssertions.java

```diff
@@ -287,140 +292,86 @@ public Boolean call() {
-                    if (null != actual && actual.toString().matches(expected.toString())) {
-                        abort(target, condition, expected, actual);
-                    }
-                    break;
-
-                case HAS_TRUTH_VALUE :
-                    if (isTrue(expected) == isTrue(actual)) {
-                        abort(target, condition, expected, actual);
-                    }
-                    break;
-
-                default:
-                    abort(target, condition, expected, actual);
+            if (!knownCondition(condition)) {
+                abort(target, UNKNOWN_CONDITION, expected, actual);
+            } else if (conditionHolds(condition, actual, expected)) {
+                abort(target, condition, expected, actual);
             }
         }
     }
-    
+
+    private static boolean conditionHolds(String condition, Object actual, Object expected) {
+        switch (condition) {
+        case IS_EQUAL_TO:
+        case EQUAL_TO:
+        case EQUALS:
+            return null != actual && actual.equals(expected);
+        case NOT_EQUAL:
+            return !Objects.equals(actual, expected);
+        case IS_NULL:
+            return isTrue(expected) == (null == actual);
+        case NOT_NULL:
+            return isTrue(expected) == (null != actual);
+        case CONTAINS:
+            return null != actual && actual.toString().contains(expected.toString());
+        case IS_EMPTY:
+            return isTrue(expected) == (null == actual || Strings.isEmpty(actual.toString()));
+        case NOT_EMPTY:
+            return isTrue(expected) == ((null != actual && Strings.isNonEmpty(actual.toString())));
+        case MATCHES:
+            return null != actual && actual.toString().matches(expected.toString());
+        case HAS_TRUTH_VALUE:
+            return isTrue(expected) == isTrue(actual);
+        case GREATER_THAN:
+            return canCompare(actual, expected) && compare(actual, expected) > 0;
+        case LESS_THAN:
+            return canCompare(actual, expected) && compare(actual, expected) < 0;
+        default:
+            return false;
+        }
+    }
+
+    private static boolean knownCondition(String condition) {
+        // The conditions should really be an enum!
+        Set<String> allConditions = ImmutableSet.of(
+                IS_NULL, NOT_NULL, IS_EQUAL_TO, EQUAL_TO, EQUALS, NOT_EQUAL,
+                MATCHES, CONTAINS, IS_EMPTY, NOT_EMPTY, HAS_TRUTH_VALUE,
+                GREATER_THAN, LESS_THAN, UNKNOWN_CONDITION);
+        return allConditions.contains(condition);
+    }
+
+    /** @return True if actual and expected are both non-null instances of {@code Comparable<T>}. */
+    private static boolean canCompare(@Nullable Object actual, @Nullable Object expected) {
+        return actual != null
+                && expected != null
+                && actual instanceof Comparable
+                && actual.getClass().equals(expected.getClass());
```

Comment:
I wonder if we should do any coercion, or perhaps even just give it a go to compare! For example, comparing `(Integer)1` with `(Long)2` might well be possible. However, on balance I think your code is right as it is the simplest (and avoid `ClassCastException`).